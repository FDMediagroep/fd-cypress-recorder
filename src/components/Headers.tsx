import React, { useEffect, useState } from 'react';
import HeadersStore = require('../stores/HeadersStore');
import { Header } from '../utils/FdEvents';
import styles from './Headers.module.scss';
import { TextInput } from '@fdmg/design-system/components/input/TextInput';

interface Props {
    headers?: Header[];
}

/**
 * Layout of the HTTP Headers tab in the Chrome Plugin UI
 */
// export default class Headers extends ComponentBase<any, State> {
export default function Headers(props: Props) {
    const [tableData, setTableData] = useState(
        HeadersStore.getHeaders() ??
            props.headers ?? [{ property: '', value: '' }]
    );

    useEffect(() => {
        const headerId = HeadersStore.subscribe(() => {
            const headers = HeadersStore.getHeaders() ?? props.headers;
            if (headers.length === 0) {
                headers.push({ property: '', value: '' });
            }
            setTableData(headers);
        });

        const headers = HeadersStore.getHeaders() ?? props.headers;
        if (headers.length === 0) {
            headers.push({ property: '', value: '' });
        }
        setTableData(headers);

        return () => {
            HeadersStore.unsubscribe(headerId);
        };
    }, []);

    const removeMultipleEmptyRows = (tblData: Header[]) => {
        return tblData.reduce((prev: Header[], header) => {
            const found = prev.find(
                (h: Header) =>
                    h.property === header.property && h.value === header.value
            );
            if (!found || header.property !== '' || header.value !== '') {
                prev.push(header);
            }
            return prev;
        }, []);
    };

    const handleHeaderChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const changedCol = e.currentTarget.getAttribute('data-column');
        const changedRow = parseInt(
            e.currentTarget.getAttribute('data-row') || '-1',
            10
        );
        let tblData: Header[] = [...tableData];
        let addNewRow = true;
        tblData =
            tblData.map<any>((data: Header, rowIndex) => {
                if (rowIndex === changedRow) {
                    switch (changedCol) {
                        case 'property':
                            return { ...data, property: e.currentTarget.value };
                        case 'value':
                            return { ...data, value: e.currentTarget.value };
                    }
                } else {
                    return { ...data };
                }
                if (!data.property && !data.value) {
                    addNewRow = false;
                }
            }) || [];
        if (addNewRow) {
            tblData.push({ property: '', value: '' });
        }
        tblData = removeMultipleEmptyRows(tblData);
        HeadersStore.setHeaders(tblData);
    };

    return (
        <table>
            <thead>
                <tr>
                    <td>Property</td>
                    <td>Value</td>
                </tr>
            </thead>
            <tbody>
                {tableData.map((data, rowIndex) => (
                    <tr key={rowIndex}>
                        <td>
                            <TextInput
                                id={`${data.property}-${data.value}`}
                                type="text"
                                className={styles.input}
                                defaultValue={data.property}
                                data-row={rowIndex}
                                data-column="property"
                                onChange={handleHeaderChange}
                            />
                        </td>
                        <td>
                            <TextInput
                                id={`${data.property}-${data.value}`}
                                type="text"
                                className={styles.input}
                                defaultValue={data.value}
                                data-row={rowIndex}
                                data-column="value"
                                onChange={handleHeaderChange}
                            />
                        </td>
                    </tr>
                ))}
            </tbody>
        </table>
    );
}
