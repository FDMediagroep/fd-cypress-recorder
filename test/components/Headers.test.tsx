import React from 'react';
import TestRenderer from 'react-test-renderer';
import Headers from '../../src/components/Headers';

describe('Headers', () => {
    it('should render correctly', () => {
        let headers = TestRenderer.create(<Headers headers={[]} />);
        expect(headers.toJSON()).toMatchSnapshot();
    });
});
