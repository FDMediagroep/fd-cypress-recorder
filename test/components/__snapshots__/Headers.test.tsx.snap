// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Headers should render correctly 1`] = `
<table>
  <thead>
    <tr>
      <td>
        Property
      </td>
      <td>
        Value
      </td>
    </tr>
  </thead>
  <tr>
    <td>
      <input
        className="sc-bdVaJa gZdsYr property"
        data-column="property"
        data-row={0}
        defaultValue=""
        onChange={[Function]}
        type="text"
      />
    </td>
    <td>
      <input
        className="sc-bdVaJa gZdsYr value"
        data-column="value"
        data-row={0}
        defaultValue=""
        onChange={[Function]}
        type="text"
      />
    </td>
  </tr>
</table>
`;
